### PR TITLE
[MPS] call contiguous after permute

### PR DESCRIPTION
**Pull request by kashif**
Fixes for MPS device.

I believe is a better fix that #1410 

Testing with on main branch:

```py
from diffusers import StableDiffusionPipeline, EulerDiscreteScheduler


model_id = "stabilityai/stable-diffusion-2"
scheduler = EulerDiscreteScheduler.from_pretrained(model_id, subfolder="scheduler")
pipe = StableDiffusionPipeline.from_pretrained(model_id, scheduler=scheduler).to("mps")

prompt = "beautiful gaze"
image = pipe(prompt, height=768, width=768).images[0]    
```

### CODE REVIEW

**Comment by kashif** [comment]
@pcuenca I can also fix the warning:

```
diffusers/src/diffusers/schedulers/scheduling_euler_discrete.py:128: UserWarning: The operator 'aten::nonzero' is not currently supported on the MPS backend and will fall back to run on the CPU. This may have performance implications. (Triggered internally at /Users/runner/work/pytorch/pytorch/pytorch/aten/src/ATen/mps/MPSFallback.mm:11.)
  step_index = (self.timesteps == timestep).nonzero().item()
```

By replacing it with:

```py
step_index = self.timesteps.tolist().index(timestep)
```

what do you think?

**Comment by pcuenca** [comment]
> @pcuenca I can also fix the warning:
> 
> ```
> diffusers/src/diffusers/schedulers/scheduling_euler_discrete.py:128: UserWarning: The operator 'aten::nonzero' is not currently supported on the MPS backend and will fall back to run on the CPU. This may have performance implications. (Triggered internally at /Users/runner/work/pytorch/pytorch/pytorch/aten/src/ATen/mps/MPSFallback.mm:11.)
>   step_index = (self.timesteps == timestep).nonzero().item()
> ```
> 
> By replacing it with:
> 
> ```python
> step_index = self.timesteps.tolist().index(timestep)
> ```
> 
> what do you think?

I think it should be ok, let's do it and see what other people think.

It's nice to remove the warning, but that operation will still move the tensor to the CPU (just as the fallback implementation of `nonzero` would do).

File: src/diffusers/models/attention.py [hunk 1]
```diff
@@ -221,11 +221,15 @@ def forward(self, hidden_states, encoder_hidden_states=None, timestep=None, retu
         # 3. Output
         if self.is_input_continuous:
             if not self.use_linear_projection:
-                hidden_states = hidden_states.reshape(batch, height, weight, inner_dim).permute(0, 3, 1, 2)
+                hidden_states = (
+                    hidden_states.reshape(batch, height, weight, inner_dim).permute(0, 3, 1, 2).contiguous()
```
Comment (pcuenca) [inline src/diffusers/models/attention.py:225]: Shouldn't we do the contiguous on `mps` only?

**Review by pcuenca — APPROVED** [review]
Thanks a lot for working on this!

File: src/diffusers/models/attention.py (hunk 1 above)
Comment (kashif) [reply src/diffusers/models/attention.py:225]: I believe everywhere else where we reshape and permute we also make contiguous... so it follows that convention

Comment (kashif) [reply src/diffusers/models/attention.py:225]: In particular in the VQ-VAE and VAE models...

**Comment by kashif** [comment]
@pcuenca at the moment with this scheduler with `fp16` on mps is also working for me:

```py
pipe = StableDiffusionPipeline.from_pretrained(
    model_id,
    torch_dtype=torch.float16,
    revision="fp16",
    scheduler=scheduler
).to("mps")
```

**Comment by patrickvonplaten** [comment]
@pcuenca I leave it up to you to merge :-)

**Review by pcuenca — APPROVED** [review]

**Comment by kashif** [comment]
@pcuenca let me double check this now on my cuda box

**Comment by kashif** [comment]
@pcuenca also works fine on "cuda"

File: src/diffusers/schedulers/scheduling_euler_ancestral_discrete.py [hunk 2]
```diff
@@ -121,7 +121,7 @@ def scale_model_input(
         """
         if isinstance(timestep, torch.Tensor):
             timestep = timestep.to(self.timesteps.device)
-        step_index = (self.timesteps == timestep).nonzero().item()
+        step_index = self.timesteps.tolist().index(timestep)
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_euler_ancestral_discrete.py:124]: why do we need this?

Comment (patrickvonplaten) [reply src/diffusers/schedulers/scheduling_euler_ancestral_discrete.py:124]: Is this needed to make it work with MPS?

**Pull request merged**